### DISCUSSION TITLE
How to use native tab navigation?

### DISCUSSION THREAD

**Discussion by fabianeichinger** [Instawork/hyperview#781] [General]
The [hv-navigator](https://github.com/Instawork/hyperview/tree/master/src/core/components/hv-navigator) module seems to provide "native" tab navigation with react-navigation, but I was unable to figure out how to actually use it. Would it be possible to get a HXML example?

**Comment by adamstep** [comment]
@hgray-instawork , any pointers? This is a new navigation framework we just finished, but the docs haven't caught up yet!

**Comment by hgray-instawork** [comment]
I apologize, @fabianeichinger for the lack of updated documentation. We have been busy adding several useful new features and haven't fully documented the changes. Rest assured that we'll be updating documents and examples in the near future. In the meantime, a basic tab navigator can be set up by defining an HXML structure like this:

```XML
<doc xmlns="https://hyperview.org/hyperview">
  <navigator id="root-navigator" type="tab">
    <nav-route id="tab-1" href="/tab1.xml" />
    <nav-route id="tab-2" href="/tab2.xml" />
    <nav-route id="tab-3" href="/tab3.xml" />
  </navigator>
</doc>
```

You can then navigate to each tab using a behavior with a `navigate` action and an href directed at the intended tab:
```XML
<behavior action="navigate" href="#tab-2" />
```

**Reply by fabianeichinger** [reply]
Thanks for replying! I tried your suggested HXML in the demo iOS Expo app but I'm not getting a tab bar or any of the tabs to show. The app is not requesting any of the tab hrefs from my backend. Any idea what I could be missing?

**Reply by hgray-instawork** [reply]
@fabianeichinger, I would expect the contents of the first tab to show up on load as a tab bar always has one selected item. That is either the first defined tab, or the first route marked with a `selected="true"` attribute. Is it possible your file paths are incorrect?

Regarding the tabs, our implementation does not include a UI. You will need to implement your own. You can do this using the tabs implementation explained on https://hyperview.org/docs/example_tabs.

**Reply by adamstep** [reply]
@hgray-instawork it would be nice to have a full end-to-end example of using the new navigation system with a bottom tab implementation like we do in the IW apps. Can you add an issue for us to add that to the website?

**Reply by hgray-instawork** [reply]
Absolutely @adamstep, we have been capturing a list of missing documentation and examples and are planning to attack these soon.

**Reply by fabianeichinger** [reply]
@hgray-instawork I set up a route to return your `<doc>` HXML example. The app is loading the route but the server logs are not showing requests for any tab href url. Also not sure where I'd place the bottom tab UI in my HXML (the schema seems to disallow `<screen>` to be a sibling of `<navigator>`) so I might misunderstand how it works conceptually. A full example would be much appreciated.

**Reply by hgray-instawork** [reply]
@fabianeichinger, to implement the UI, you should create an instance of the tab bar at the bottom of each page. Example "tab1.xml" would look something like this:
```xml
<doc ...>
   <screen ...>
      <body>
         <!-- main view content here -->
         <!-- tab bar goes here -->
      </body>
   </screen>
</doc>
```

**Reply by hgray-instawork** [reply]
Regarding your issue of loading, I'm not quite sure what could be going on. You should be passing the url to your navigation document as the `entrypointUrl` value in Hyperview. Assuming your navigation structure is in a file called `index.xml`, the Hyperview would be instantiated as

```Javascript
return (
    <Hyperview
      ...
      entrypointUrl='/index.xml'
      ...
    />
);
```

**Reply by hgray-instawork** [reply]
@fabianeichinger , were you able to resolve your issues? We have recently released v0.75.0 of Hyperview which now includes an example of using a navigator hierarchy provided by an HXML document. You can find this in the `demo` directory included in the Hyperview source.

Please let us know how you are progressing and if the new example is helpful to you.

**Comment by hgray-instawork** [comment]
@fabianeichinger , I hope everything is going well with your navigation implementation. I wanted to let you know that we've just updated our documentation to include a more in-depth description of the navigation functionality, settings, and setup. Please have a look at our new [Navigation guide](https://hyperview.org/docs/guide_navigation) and let us know how things are going!
